src/distances/distance_metric.rs

```diff
@@ -362,3 +377,416 @@ impl LvEditDistanceMetric {
+                if g1 == g2 {
+                    let idx_prev = Self::idx_at(i - 1, j - 1, len_w2);
+                    self.dplv[idx_cur] = self.dplv[idx_prev];
+                    self.dpss[idx_cur] = self.dpss[idx_prev] + 1;
+                    longest = max(longest, self.dpss[idx_cur]);
+                    continue;
+                }
+
+                let idx_sub = Self::idx_at(i - 1, j - 1, len_w2);
+                let idx_del = Self::idx_at(i - 1, j, len_w2);
+                let idx_add = Self::idx_at(i, j - 1, len_w2);
+
+                let len_sub = self.dplv[idx_sub];
+                let len_del = self.dplv[idx_del];
+                let len_add = self.dplv[idx_add];
+
+                if len_sub < len_del && len_sub < len_add {
+                    self.dplv[idx_cur] = len_sub + 1;
+                } else if len_del < len_add {
+                    self.dplv[idx_cur] = len_del + 1;
+                } else {
+                    self.dplv[idx_cur] = len_add + 1;
+                }
+            }
+        }
+
+        let idx = Self::idx_at(len_w1, len_w2, len_w2);
+        (self.dplv[idx], longest)
+    }
+}
+
+impl DistanceMetric<Word> for LvSubstringDistanceMetric {
+    fn dist(&mut self, v1: &Word, v2: &Word) -> f32 {
+        let (edits, longest_substring) = self.compute_edits(v1, v2);
+        let bonus = longest_substring as f32 * self.weight;
+        let edits = f32::max(edits as f32 - bonus, 0.0);
+        1.0 - edits / usize::max(v1.raw.len(), v2.raw.len()) as f32
+    }
+}
+
+pub struct LvMultiWordDistanceMetric {
+    lvopti: LvOptiDistanceMetric,
+    separator: GraphemeType,
+}
+
+impl LvMultiWordDistanceMetric {
+    pub fn new(separator: GraphemeType) -> Self {
+        Self {
+            lvopti: LvOptiDistanceMetric::new(),
+            separator,
+        }
+    }
+
+    fn extract_multiple_words(w: &Word, separator: GraphemeType) -> Vec<Vec<GraphemeType>> {
+        let mut ret: Vec<Vec<GraphemeType>> = Vec::new();
+        let mut temp: Vec<GraphemeType> = Vec::new();
+
+        for g in &w.graphemes {
+            if *g == separator {
+                if !temp.is_empty() {
+                    ret.push(temp);
+                }
+                temp = Vec::new();
+            } else {
+                temp.push(*g);
+            }
+        }
+        if !temp.is_empty() {
+            ret.push(temp);
+        }
+        ret
+    }
+
+    fn compute_edits(&mut self, w1: &Word, w2: &Word) -> u8 {
+        let multi_w1 = Self::extract_multiple_words(w1, self.separator);
+        let multi_w2 = Self::extract_multiple_words(w2, self.separator);
+
+        let mut total = 0;
+        let mut perfect_matches = 0;
+
+        let len1 = multi_w1.len();
+        let len2 = multi_w2.len();
+
+        if len1 > 1 && len1 == len2 {
+            for i in 0..len1 {
+                let w1 = &Word::from_graphemes(multi_w1[i].clone());
+                let w2 = &Word::from_graphemes(multi_w2[i].clone());
```

Comment:
The current implementation (especially these lines) is not very efficient, so if this metric turns out to be the best suited one will have to refactor a bit.